### PR TITLE
Add voter reward commission for validator groups

### PR DESCRIPTION
**Pull request by pahor167**
## Summary

Introduces a voter reward commission mechanism that allows validator groups to take a configurable percentage of voter CELO epoch rewards. This is separate from the existing commission field which governs the group's cut of validator cUSD payments.

**Example:** If a group sets 10% voter reward commission and voters for that group earn 1,000 CELO in epoch rewards, the group receives 100 CELO (released from CeloUnreleasedTreasury) and voters receive the remaining 900 CELO as vote credit inflation.

### EpochManager.sol
- `_deductVoterRewardCommission()` — deducts commission from voter rewards and releases CELO from treasury to the group
- Commission is **clamped** to `maxVoterRewardCommission` at distribution time, so governance cap changes take effect immediately for all groups (even those with previously-activated higher rates)
- Integrated into both `processGroup()` and `finishNextEpochProcess()` paths
- New `VoterRewardCommissionDistributed` event

### Validators.sol
- New `voterRewardCommission`, `nextVoterRewardCommission`, `nextVoterRewardCommissionBlock` fields on `ValidatorGroup` struct
- `setNextVoterRewardCommissionUpdate()` / `updateVoterRewardCommission()` — time-delayed queue/activate pattern matching existing `commission` flow, reusing `commissionUpdateDelay`
- `setMaxVoterRewardCommission()` — governance-controlled cap (0 = no cap)
- Max cap re-checked at activation time to handle governance lowering the cap after queuing
- `getVoterRewardCommission()` — returns current, pending, and activation block

### Economics
- Commission is carved from the already-budgeted `totalRewardsVoter` pool — it redirects part of the voter reward from deferred LockedGold claims (vote credit inflation) to immediate treasury releases
- The total economic cost per epoch is unchanged; commission changes the distribution mechanism, not the total emission

### Tests
- Unit tests covering: 0% commission, 10%, 100%, rounding-to-zero, sentinel values, both processing paths, event emission, queue/activate lifecycle, delay enforcement, max cap enforcement, cap-lowered-after-queue
- **Fuzz tests**: conservation invariant (`commission + voterRewards == totalRewards`) across random commission rates and reward amounts, math verification against FixidityLib calculation
- **Epoch-timing tests**: proving commission uses the rate active at `processGroup()` time, conservation holds when commission changes mid-processing
- **Max cap clamp tests**: clamping at distribution time, no-clamp below cap, no-clamp when cap is 0, fuzz conservation with clamping

### Storage Safety
- `ValidatorGroup` struct fields appended at end (safe for mapping-based storage)
- `maxVoterRewardCommission` state variable appended after `deprecated_downtimeGracePeriod` (safe for proxy upgrade)

## Deployment Note
After upgrade, `maxVoterRewardCommission` defaults to 0 (no cap). The governance proposal should atomically call `setMaxVoterRewardCommission()` to set the desired cap.

### CODE REVIEW

**Comment by pahor167** [comment]
@codex review

**Comment by pahor167** [comment]
## review'd by Claude

### 🟡 Suggestion (3)

- **EpochManager version number not bumped despite new behavior** — `packages/protocol/contracts-0.8/common/EpochManager.sol` (line 626)
  The EpochManager version remains (1,1,0,3) — same as the base branch — despite adding new internal logic (_deductVoterRewardCommission), a new event (VoterRewardCommissionDistributed), and additional treasury releases. The second commit (1ad36689d) reverted the version bump from (1,1,1,0) back to (1,1,0,3) with the rationale 'internal logic change = patch bump', but then didn't actually apply the patch bump. Even for internal changes, a patch bump (e.g., 1,1,0,4) is warranted since the contract's observable behavior changes (new event, different reward distribution, additional treasury releases).
- **updateVoterRewardCommission does not re-check maxVoterRewardCommission** — `packages/protocol/contracts-0.8/governance/Validators.sol` (line 525)
  When applying a queued voter reward commission update, the function does not re-validate against `maxVoterRewardCommission`. If governance lowers the cap between queue and activation (e.g., group queues 50%, governance sets max to 20%), the group can still activate the 50% commission. While this mirrors the existing `updateCommission()` pattern (which also doesn't re-check), the voter reward commission has a treasury cost impact that the regular commission doesn't — it directly releases CELO from treasury. Consider re-checking the cap at activation time, or at minimum document this as a known behavior.
- **Commission releases additional CELO from treasury beyond calculated epoch rewards** — `packages/protocol/contracts-0.8/common/EpochManager.sol` (line 730)
  The voter reward commission is paid as additional CELO from CeloUnreleasedTreasury, beyond the amounts calculated by EpochRewards.calculateTargetEpochRewards(). Voter rewards are 'virtual' (they increase vote weight via Election.distributeEpochRewards without releasing CELO), but the commission converts a portion into real CELO. This means with e.g., 10% commission across all groups, the treasury releases an additional ~10% of totalRewardsVoter per epoch. While the code comments suggest this is intentional, the EpochRewards target calculations don't account for this additional inflation. If this is by design, consider adding a NatSpec comment on `_deductVoterRewardCommission` explaining this economic trade-off, since future maintainers will question it.

### 🔵 Nitpick (1)

- **Inconsistent variable naming between processGroup and finishNextEpochProcess** — `packages/protocol/contracts-0.8/common/EpochManager.sol` (line 330)
  In `processGroup()`, the commission deduction uses explicit variable names (`commissionAmount`, `voterRewards`), while in `finishNextEpochProcess()` it reuses the `epochRewards` variable in-place (`epochRewards -= _deductVoterRewardCommission(...)`). Both approaches work correctly, but the explicit naming in processGroup is clearer. Consider using the same explicit pattern in finishNextEpochProcess for consistency and readability.

### 🟢 Good (2)

- **Clean implementation following established patterns** — `packages/protocol/contracts-0.8/governance/Validators.sol` (line 478)
  The queue/update pattern for voter reward commission closely mirrors the existing commission update flow, including the commission update delay, the `_sendValidatorGroupPaymentsIfNecessary` call before applying, and the max cap validation. This consistency makes the code predictable and easy to audit.
- **Thorough test coverage including edge cases** — `packages/protocol/test-sol/unit/common/EpochManager.t.sol` (line 990)
  The tests cover important edge cases: zero commission, full (100%) commission, commission that rounds to zero, the sentinel value (type(uint256).max for 0 rewards), and both processGroup and finishNextEpochProcess code paths. The Validators tests also cover revert cases, overwriting queued updates, and max commission enforcement.

**Bottom line:** 1. **Bump EpochManager version** — currently unchanged from base, should be at least (1,1,0,4). 2. **Consider re-checking maxVoterRewardCommission at activation** — the commission has direct treasury cost unlike the existing commission, so bypassing the cap is more consequential. 3. **Document the economic model** — the treasury releases additional CELO beyond the EpochRewards target calculations; this should be explicitly documented for future maintainers.

*Automated review by [reviewd](https://github.com/simion/reviewd). Findings are AI-generated — use your judgment.*
*Replies to this comment are not monitored.*

[](reviewd)

File: packages/protocol/contracts-0.8/common/EpochManager.sol [hunk 1]
```diff
@@ -682,6 +701,37 @@ contract EpochManager is
...
+      .newFixed(epochRewards)
+      .multiply(FixidityLib.wrap(voterRewardCommissionUnwrapped))
+      .fromFixed();
+
+    if (commissionAmount > 0) {
+      // Release CELO from treasury directly to the group.
+      // This mirrors the pattern used for community and carbon fund rewards
+      // in _finishEpochHelper().
+      getCeloUnreleasedTreasury().release(group, commissionAmount);
+      emit VoterRewardCommissionDistributed(group, commissionAmount, currentEpochNumber);
```
[the remark this answers is not in the document]
Comment (pahor167) [reply packages/protocol/contracts-0.8/common/EpochManager.sol:746]: Addressed — added NatSpec on `_deductVoterRewardCommission` documenting this is intentional and explaining the treasury outflow trade-off.

File: packages/protocol/contracts-0.8/governance/Validators.sol [hunk 2]
```diff
@@ -452,6 +469,65 @@ contract Validators is
...
+   * update.
+   * @dev Note: Consider adding onlyWhenNotBlocked modifier if Validators inherits Blockable
+   * in the future, to prevent updates during epoch processing.
+   */
+  function updateVoterRewardCommission() external {
+    address account = getAccounts().validatorSignerToAccount(msg.sender);
+    require(isValidatorGroup(account), "Not a validator group");
+    ValidatorGroup storage group = groups[account];
+
+    _sendValidatorGroupPaymentsIfNecessary(group);
```
[the remark this answers is not in the document]
Comment (pahor167) [reply packages/protocol/contracts-0.8/governance/Validators.sol]: Fixed — removed the unnecessary `_sendValidatorGroupPaymentsIfNecessary` call since `voterRewardCommission` is independent from cUSD payment commission.

File: packages/protocol/contracts-0.8/governance/Validators.sol [hunk 3]
```diff
@@ -452,6 +469,65 @@ contract Validators is
...
+
+    require(group.nextVoterRewardCommissionBlock != 0, "No voter reward commission update queued");
+    require(
+      group.nextVoterRewardCommissionBlock <= block.number,
+      "Can't apply voter reward commission update yet"
+    );
+
+    group.voterRewardCommission = group.nextVoterRewardCommission;
```
[the remark this answers is not in the document]
Comment (pahor167) [reply packages/protocol/contracts-0.8/governance/Validators.sol:530]: Fixed — `updateVoterRewardCommission` now re-checks `maxVoterRewardCommission` at activation time.

**Comment by pahor167** [comment]
Addressed in d2563cac7:
- **Version bump**: bumped to (1,1,0,4) ✅
- **Re-check maxVoterRewardCommission**: added at activation time with tests ✅
- **Treasury docs**: added NatSpec explaining the economic trade-off ✅
- **Naming inconsistency**: can't use explicit vars due to stack-too-deep, added comment explaining why ✅

File: packages/protocol/contracts-0.8/governance/Validators.sol [hunk 4]
```diff
@@ -452,6 +469,65 @@ contract Validators is
...
+    );
+  }
+
+  /**
+   * @notice Updates a validator group's voter reward commission based on the previously queued
+   * update.
+   * @dev Note: Consider adding onlyWhenNotBlocked modifier if Validators inherits Blockable
+   * in the future, to prevent updates during epoch processing.
+   */
+  function updateVoterRewardCommission() external {
```
[the remark this answers is not in the document]
Comment (pahor167) [reply packages/protocol/contracts-0.8/governance/Validators.sol:510]: Not valid since there is waiting window > Epoch

File: packages/protocol/contracts-0.8/common/EpochManager.sol [hunk 5]
```diff
@@ -682,6 +703,52 @@ contract EpochManager is
...
+    IValidators validators = getValidators();
+    (uint256 voterRewardCommissionUnwrapped, , ) = validators.getVoterRewardCommission(group);
+
+    if (voterRewardCommissionUnwrapped == 0) {
+      return 0;
+    }
+
+    // Clamp to the governance-set max cap so that previously-activated commissions
+    // exceeding a later-lowered cap are still bounded at distribution time.
+    uint256 maxCommission = validators.maxVoterRewardCommission();
```
[the remark this answers is not in the document]
Comment (pahor167) [reply packages/protocol/contracts-0.8/common/EpochManager.sol:731]: setMaxVoterRewardCommission is delayed

File: packages/protocol/contracts-0.8/governance/Validators.sol [hunk 6]
```diff
@@ -129,6 +134,11 @@ contract Validators is
   uint256 public slashingMultiplierResetPeriod;
   uint256 public deprecated_downtimeGracePeriod;
 
+  // Cap on voter reward commission to protect voters from excessive commission rates.
+  // Set via governance. A value of 0 means no cap is enforced.
+  // FixidityLib representation
```
Comment (m-chrzan) [inline packages/protocol/contracts-0.8/governance/Validators.sol:139]: Should this be a FixidityLib.Fraction value to make this clear, instead of using a comment?

File: packages/protocol/contracts-0.8/governance/Validators.sol [hunk 7]
```diff
@@ -452,6 +469,70 @@ contract Validators is
...
+        commission <= maxVoterRewardCommission,
+        "Voter reward commission exceeds max allowed"
+      );
+    }
+    require(
+      commission != group.voterRewardCommission.unwrap(),
+      "Voter reward commission must be different"
+    );
+
+    group.nextVoterRewardCommission = FixidityLib.wrap(commission);
```
Comment (m-chrzan) [inline packages/protocol/contracts-0.8/governance/Validators.sol:497]: Suggestion: wrap into Fixidity earlier in function, then the comparisons in checks above don't need `.unwrap`.

File: packages/protocol/contracts-0.8/governance/Validators.sol [hunk 8]
```diff
@@ -452,6 +469,70 @@ contract Validators is
...
+    group.nextVoterRewardCommissionBlock = block.number.add(commissionUpdateDelay);
```
Comment (m-chrzan) [inline packages/protocol/contracts-0.8/governance/Validators.sol:498]: Are we still using SafeMath to keep with old conventions, or should we use 0.8's default checked arithmetic in new logic?

File: packages/protocol/contracts-0.8/governance/Validators.sol [hunk 9]
```diff
@@ -821,6 +921,22 @@ contract Validators is
     emit CommissionUpdateDelaySet(delay);
   }
 
+  /**
+   * @notice Sets the maximum voter reward commission that groups can set.
+   * @param maxCommission Fixidity representation of the max commission.
+   *   A value of 0 means no cap is enforced.
```
Comment (m-chrzan) [inline packages/protocol/contracts-0.8/governance/Validators.sol:927]: Wondering if this is the right approach. Having no special value would be conceptually simpler and
1. Simplify some of the logic above (no need for `if (maxCommission > 0)` blocks)
2. Still easily allow an "unlimited" setting (by setting maxCommission to fixed1)
3. Start the system with reward commissions turned off, and require an explicit Governance decision to set the initial value.

File: packages/protocol/test-sol/unit/common/EpochManager.t.sol [hunk 10]
```diff
@@ -985,3 +990,562 @@ contract EpochManagerTest_getElectedSignerByIndex is EpochManagerTest {
...
+
+    epochManagerContract.startNextEpochProcess();
+    epochManagerContract.setToProcessGroups();
+
+    uint256 groupBalanceBefore = celoToken.balanceOf(group);
+
+    epochManagerContract.processGroup(group, address(0), address(0));
+
+    // Voters should receive 90% of rewards
+    uint256 expectedVoterRewards = 900e18;
```
Comment (m-chrzan) [inline packages/protocol/test-sol/unit/common/EpochManager.t.sol:1029]: `900 ether` instead of `e18`?

File: packages/protocol/test-sol/unit/common/EpochManager.t.sol [hunk 11]
```diff
@@ -985,3 +990,562 @@ contract EpochManagerTest_getElectedSignerByIndex is EpochManagerTest {
...
+    assertEq(
+      election.distributedEpochRewards(group),
+      expectedVoterRewards,
+      "Voters should receive rewards minus commission"
+    );
+
+    // Group should receive 10% as CELO from treasury
+    uint256 expectedCommission = 100e18;
```
Comment (m-chrzan) [inline packages/protocol/test-sol/unit/common/EpochManager.t.sol:1037]: ditto

File: packages/protocol/test-sol/unit/common/EpochManager.t.sol [hunk 12]
```diff
@@ -985,3 +990,562 @@ contract EpochManagerTest_getElectedSignerByIndex is EpochManagerTest {
...
+    vm.expectEmit(true, true, true, true);
+    emit VoterRewardCommissionDistributed(group, expectedCommission, firstEpochNumber);
+
+    epochManagerContract.processGroup(group, address(0), address(0));
+  }
+
+  function test_handlesZeroEpochRewardsSentinel() public {
+    // When epochRewards == type(uint256).max (sentinel for 0 rewards),
+    // no distribution should happen
+    election.setGroupEpochRewardsBasedOnScore(group, 0);
```
Comment (m-chrzan) [inline packages/protocol/test-sol/unit/common/EpochManager.t.sol:1063]: Nit: the uint256.max sentinel logic happens behind the scenes, is not referenced here. So the comment and test case name feel out of place (in the function call we pass 0 instead of the mentioned max int).

File: packages/protocol/test-sol/unit/common/EpochManager.t.sol [hunk 13]
```diff
@@ -985,3 +990,562 @@ contract EpochManagerTest_getElectedSignerByIndex is EpochManagerTest {
...
+    assertEq(
+      groupBalanceAfter - groupBalanceBefore,
+      expectedCommission,
+      "Group should receive commission via finishNextEpochProcess"
+    );
+  }
+
+  function test_noTreasuryReleaseWhenCommissionRoundsToZero() public {
+    // Set a very tiny commission (1 wei in Fixidity representation).
+    // With small epoch rewards, the multiply should round down to 0.
```
Comment (m-chrzan) [inline packages/protocol/test-sol/unit/common/EpochManager.t.sol:1139]:
```suggestion
    // With small epoch rewards, the multiplication result should round down to 0.
```

File: packages/protocol/test-sol/unit/common/EpochManager.t.sol [hunk 14]
```diff
@@ -985,3 +990,562 @@ contract EpochManagerTest_getElectedSignerByIndex is EpochManagerTest {
...
+    uint256 tinyCommission = 1;
+    validators.setVoterRewardCommission(group, tinyCommission);
+
+    // Use small rewards so tinyCommission * rewards rounds to 0
```
Comment (m-chrzan) [inline packages/protocol/test-sol/unit/common/EpochManager.t.sol:1143]: Nit: Is this comment necessary? The idea for this test was already explained in the comment above.

File: packages/protocol/contracts-0.8/governance/Validators.sol (hunk 9 above)
Comment (pahor167) [reply packages/protocol/contracts-0.8/governance/Validators.sol]: agreed

File: packages/protocol/contracts-0.8/common/EpochManager.sol [hunk 15]
```diff
@@ -682,6 +701,52 @@ contract EpochManager is
...
+   * from deferred LockedGold claims to immediate treasury releases. The per-epoch treasury outflow
+   * from commission equals the sum of (groupVoterRewards * groupCommission) across all elected
+   * groups, bounded by maxVoterRewardCommission.
+   */
+  function _deductVoterRewardCommission(
+    address group,
+    uint256 epochRewards
+  ) internal returns (uint256 commissionAmount) {
+    IValidators validators = getValidators();
+    (uint256 voterRewardCommissionUnwrapped, , ) = validators.getVoterRewardCommission(group);
```
[the remark this answers is not in the document]
Comment (pahor167) [reply packages/protocol/contracts-0.8/common/EpochManager.sol:735]: Fixed — `_deductVoterRewardCommission` now checks `validators.isValidatorGroup(group)` before calling `getVoterRewardCommission`. If the group deregistered between epoch end and reward processing, commission is skipped (returns 0) and epoch finalization proceeds normally. Added test `test_skipsCommissionWhenGroupDeregisteredBeforeProcessing`.

[the remark this answers is not in the document]
Comment (pahor167) [reply packages/protocol/contracts-0.8/common/EpochManager.sol:735]: Fixed — same as above. Added `isValidatorGroup` guard in `_deductVoterRewardCommission` so deregistered groups don't block epoch finalization.

File: packages/protocol/contracts-0.8/governance/Validators.sol (hunk 6 above)
Comment (pahor167) [reply packages/protocol/contracts-0.8/governance/Validators.sol]: Fixed — changed to `FixidityLib.Fraction public maxVoterRewardCommission`.

File: packages/protocol/contracts-0.8/governance/Validators.sol (hunk 7 above)
Comment (pahor167) [reply packages/protocol/contracts-0.8/governance/Validators.sol]: Fixed — wrapping into Fraction early and using `.lte()`, `.equals()` for comparisons instead of `.unwrap()`.

File: packages/protocol/contracts-0.8/governance/Validators.sol (hunk 8 above)
Comment (pahor167) [reply packages/protocol/contracts-0.8/governance/Validators.sol]: Using 0.8 native checked arithmetic for new code — SafeMath is redundant in 0.8 and only remains in legacy code.

File: packages/protocol/contracts-0.8/governance/Validators.sol (hunk 9 above)
Comment (pahor167) [reply packages/protocol/contracts-0.8/governance/Validators.sol]: Agreed — removed the `0 = no cap` special case. Now `maxVoterRewardCommission` defaults to 0 (commissions disabled), governance sets it to `FixidityLib.fixed1()` for unlimited. Simplified all the `if (maxVoterRewardCommission > 0)` guards away.

File: packages/protocol/test-sol/unit/common/EpochManager.t.sol (hunk 10 above)
Comment (pahor167) [reply packages/protocol/test-sol/unit/common/EpochManager.t.sol]: Fixed.

File: packages/protocol/test-sol/unit/common/EpochManager.t.sol (hunk 11 above)
Comment (pahor167) [reply packages/protocol/test-sol/unit/common/EpochManager.t.sol]: Fixed.

File: packages/protocol/test-sol/unit/common/EpochManager.t.sol (hunk 12 above)
Comment (pahor167) [reply packages/protocol/test-sol/unit/common/EpochManager.t.sol]: Fixed — renamed to `test_handlesZeroEpochRewards` and simplified comment.

File: packages/protocol/test-sol/unit/common/EpochManager.t.sol (hunk 13 above)
Comment (pahor167) [reply packages/protocol/test-sol/unit/common/EpochManager.t.sol]: Fixed.

File: packages/protocol/test-sol/unit/common/EpochManager.t.sol (hunk 14 above)
Comment (pahor167) [reply packages/protocol/test-sol/unit/common/EpochManager.t.sol]: Fixed — removed.

File: .github/workflows/protocol_tests.yml [hunk 16]
```diff
@@ -180,13 +180,34 @@ jobs:
...
+          path: celo-optimism
+          submodules: recursive
+      - name: Generate initial state for devchain
+        run: yarn devchain:generate-initial-state --contracts-dir ${{ github.workspace }}/celo-optimism/packages/contracts-bedrock
+      - name: Install celocli
+        run: sudo npm install -g @celo/celocli@latest --unsafe-perm
       - name: Generate migrations and run devchain
-        run: ./scripts/foundry/create_and_migrate_anvil_devchain.sh
+        run: |
+          LOAD_STATE=${{ github.workspace }}/celo-optimism/packages/contracts-bedrock/anvil-state.json ./scripts/foundry/create_and_migrate_anvil_devchain.sh
```
Comment (martinvol) [inline .github/workflows/protocol_tests.yml:210]: I'm puzzled why this didn't make my PR fail before https://github.com/celo-org/celo-monorepo/pull/11593

File: packages/protocol/contracts-0.8/common/EpochManager.sol [hunk 17]
```diff
@@ -320,7 +332,11 @@ contract EpochManager is
     IElection election = getElection();
 
     if (epochRewards != type(uint256).max) {
-      election.distributeEpochRewards(group, epochRewards, lesser, greater);
+      uint256 commissionAmount = _deductVoterRewardCommission(group, epochRewards);
+      uint256 voterRewards = epochRewards - commissionAmount;
```
Comment (martinvol) [inline packages/protocol/contracts-0.8/common/EpochManager.sol]: As the internal function is called `_deductVoterRewardCommission`, I'd expect this math (`epochRewards - commissionAmount`) to be inside of it and to return a tuple.

File: packages/protocol/contracts-0.8/common/EpochManager.sol [hunk 18]
```diff
@@ -687,6 +706,58 @@ contract EpochManager is
...
+  function _deductVoterRewardCommission(
+    address group,
+    uint256 epochRewards
+  ) internal returns (uint256 commissionAmount) {
+    IValidators validators = getValidators();
+
+    // If the group deregistered between epoch end and reward processing, skip commission.
+    if (!validators.isValidatorGroup(group)) {
+      return 0;
+    }
```
Comment (martinvol) [inline packages/protocol/contracts-0.8/common/EpochManager.sol:733]: Is is actually possible? I'd leave the check anyway, but add a comment if we don't expect this case to ever hit.

File: packages/protocol/contracts-0.8/common/EpochManager.sol [hunk 19]
```diff
@@ -687,6 +706,58 @@ contract EpochManager is
...
+    if (voterRewardCommissionUnwrapped == 0) {
+      return 0;
+    }
+
+    // Clamp to the governance-set max cap so that previously-activated commissions
+    // exceeding a later-lowered cap are still bounded at distribution time.
+    uint256 maxCommission = validators.maxVoterRewardCommission();
+    if (voterRewardCommissionUnwrapped > maxCommission) {
+      voterRewardCommissionUnwrapped = maxCommission;
+    }
```
Comment (martinvol) [inline packages/protocol/contracts-0.8/common/EpochManager.sol]: If we hit this edge, it can be inconsistent as this function is called twice at different blocks. So I'd either call it only once or prevent in a way having different max for voters and groups.

In practice is this very unlikely case, but just for consistency I'd try to avoid having this odd scenario.

File: packages/protocol/contracts-0.8/governance/Validators.sol [hunk 20]
```diff
@@ -66,6 +66,11 @@ contract Validators is
     // sizeHistory[i] contains the last time the group contained i members.
     uint256[] sizeHistory;
     SlashingInfo slashInfo;
+    // Commission on voter CELO rewards (independent from validator payment commission above).
+    // Groups set this to take a percentage of epoch rewards that would otherwise go to voters.
+    FixidityLib.Fraction voterRewardCommission;
+    FixidityLib.Fraction nextVoterRewardCommission;
+    uint256 nextVoterRewardCommissionBlock;
```
Comment (martinvol) [inline packages/protocol/contracts-0.8/governance/Validators.sol:77]: Would add a comment of what this variables mean

File: packages/protocol/contracts-0.8/governance/Validators.sol [hunk 21]
```diff
@@ -452,6 +469,64 @@ contract Validators is
...
+    uint256 activationBlock = block.number + commissionUpdateDelay;
+    group.nextVoterRewardCommissionBlock = activationBlock;
+    emit ValidatorGroupVoterRewardCommissionUpdateQueued(account, commission, activationBlock);
+  }
+
+  /**
+   * @notice Updates a validator group's voter reward commission based on the previously queued
+   * update.
+   */
+  function updateVoterRewardCommission() external {
```
Comment (martinvol) [inline packages/protocol/contracts-0.8/governance/Validators.sol:521]: you could also check that the EpochManager is not mid-process and get rid of the edge case

File: packages/protocol/contracts-0.8/governance/Validators.sol [hunk 22]
```diff
@@ -452,6 +469,64 @@ contract Validators is
...
+    // Re-check max cap at activation time. Governance may have lowered the cap since the
+    // update was queued.
+    require(
+      group.nextVoterRewardCommission.lte(maxVoterRewardCommission),
+      "Voter reward commission exceeds max allowed"
+    );
+
+    group.voterRewardCommission = group.nextVoterRewardCommission;
+    delete group.nextVoterRewardCommission;
+    delete group.nextVoterRewardCommissionBlock;
```
Comment (martinvol) [inline packages/protocol/contracts-0.8/governance/Validators.sol]: why delete instead of 0?
I think it's the same result, but find it more difficult to find and I don't think we use elsewhere in the repo

File: packages/protocol/contracts-0.8/common/EpochManager.sol (hunk 18 above)
Comment (pahor167) [reply packages/protocol/contracts-0.8/common/EpochManager.sol:733]: it is possible if in between epoch switching group would stop being the validator group - highly improbable but not impossible

Comment (martinvol) [reply packages/protocol/contracts-0.8/common/EpochManager.sol:733]: you're right, I though deregister had `onlyWhenNotBlocked`, but apparently wasn't a requirement back then

**Review by Mc01 — APPROVED** [review]
